src/ringbuffer.c

```diff
@@ -103,3 +104,17 @@ void ringbuffer_reset_read_pos(struct ringbuffer_t* rb) {
 void* ringbuffer_buf_ptr(struct ringbuffer_t* rb) {
    return rb->buf;
 }
+
+
+// clone everthing except buf (the buf is shared)
+ringbuffer_t* ringbuffer_clone(ringbuffer_t* rb) {
+   ringbuffer_t* crb = malloc(sizeof(ringbuffer_t));
+   crb->buf = rb->buf;
+   crb->cb = rb->cb;
+   crb->rp = rb->rp;
+   crb->wp = rb->wp;
+   crb->rcb = rb->rcb;
+   crb->saved_wp = rb->saved_wp;
+   crb->saved_rcb = rb->saved_wp;
```

Comment:
Maybe use `memcpy` instead of copying individual fields? This highlights that everything is the same and avoids issue is we expand it.